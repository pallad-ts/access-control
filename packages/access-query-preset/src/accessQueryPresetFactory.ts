import {AccessQueryPreset} from "./AccessQueryPreset";
import {InferSubjectFromDefinition, SubjectDefinition} from "./types";
import {AccessQuery} from "@pallad/access-control";

export function accessQueryPresetFactory<T extends Record<string, readonly [string, SubjectDefinition]>>(name: string, methods: T) {

	const accessQueryPreset = new AccessQueryPreset(name);
	const result: any = {
		accessQueryPreset,
	};

	for (const [method, [action, subjectDefinition]] of Object.entries(methods)) {
		accessQueryPreset.addEntry(action, subjectDefinition);
		result[method] = (subject: any, principal: any) => {
			return {action, subject: accessQueryPreset.wrapSubjectWithSubjectType(subject), principal};
		}
	}

	return result as AccessQueryPresetFactory<T>;
}


export type AccessQueryPresetFactory<T extends Record<string, readonly [string, SubjectDefinition]>> = {
	accessQueryPreset: AccessQueryPreset<MethodsToAccessQueryParamsUnion<T[keyof T]>>,
} & {
	[TKey in keyof T]: <
		TSubject extends InferSubjectFromDefinition<T[TKey][1]>,
		TPrincipal
	>(subject: TSubject, principal: TPrincipal) => AccessQuery<TPrincipal, T[TKey][0], TSubject>
}

export type MethodsToAccessQueryParamsUnion<T extends readonly [string, SubjectDefinition]> =
	T extends readonly [infer TAction extends string, infer TSubjectDefinition extends SubjectDefinition] ?
		[TAction, InferSubjectFromDefinition<TSubjectDefinition>]
		: never;
