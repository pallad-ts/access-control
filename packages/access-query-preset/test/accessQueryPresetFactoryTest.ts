import {accessQueryPresetFactory} from "@src/accessQueryPresetFactory";
import {IsExact, assert} from "conditional-type-checks";
import {AccessQueryPreset} from "@src/AccessQueryPreset";
import {ORGANIZATION_REF, organizationFactory, PRODUCT_REF, productFactory} from "./fixtures";
import {Subject} from "@pallad/access-control";
import {PRINCIPAL} from "@pallad/access-control-dev/test/fixtures";

describe('accessQueryPresetFactory', () => {
	const FOO_BAR = {foo: 'bar'} as const;
	const factory = accessQueryPresetFactory('product', c => ({
		canUpdate: c('update', productFactory),
		canDelete: c('delete', productFactory),
		canCreate: c('create', organizationFactory),
		canTruncate: c('truncate', Subject.Global),
		canActivate: c('activate', (value: unknown): value is typeof FOO_BAR => {
			return value === FOO_BAR;
		})
	}));

	it('provided methods are assigned to final object and those methods return access control query params with action and subject', () => {
		expect(factory.canDelete(PRODUCT_REF, PRINCIPAL))
			.toMatchSnapshot();

		expect(factory.canCreate(ORGANIZATION_REF, PRINCIPAL))
			.toMatchSnapshot();

		expect(factory.canUpdate(PRODUCT_REF, PRINCIPAL))
			.toMatchSnapshot();
	});

	it('types', () => {
		type ProductEntityRef = ReturnType<typeof productFactory>;
		type OrganizationEntityRef = ReturnType<typeof organizationFactory>;

		type Factory = typeof factory;

		assert<IsExact<Parameters<Factory['canCreate']>[0], OrganizationEntityRef>>(true);
		assert<IsExact<Parameters<Factory['canDelete']>[0], ProductEntityRef>>(true);
		assert<IsExact<Parameters<Factory['canUpdate']>[0], ProductEntityRef>>(true);

		assert<IsExact<Factory['accessQueryPreset'], AccessQueryPreset<
			['update', ProductEntityRef] | ['delete', ProductEntityRef] | ['create', OrganizationEntityRef] | ['truncate', Subject.Global] | ['activate', typeof FOO_BAR]
		>>>(true);
	});

	it('returned permission set is preconfigured with all methods', () => {
		expect(factory.accessQueryPreset.canHandleQuery(
			factory.canUpdate(PRODUCT_REF, PRINCIPAL),
		)).toBe(true);

		expect(factory.accessQueryPreset.canHandleQuery(
			factory.canDelete(PRODUCT_REF, PRINCIPAL)
		)).toBe(true);

		expect(factory.accessQueryPreset.canHandleQuery(
			factory.canCreate(ORGANIZATION_REF, PRINCIPAL),
		)).toBe(true);

		expect(factory.accessQueryPreset.canHandleQuery(
			factory.canTruncate(Subject.Global.INSTANCE, PRINCIPAL)
		)).toBe(true);

		expect(factory.accessQueryPreset.canHandleQuery(
			factory.canActivate(FOO_BAR, PRINCIPAL)
		)).toBe(true);
	});
});
