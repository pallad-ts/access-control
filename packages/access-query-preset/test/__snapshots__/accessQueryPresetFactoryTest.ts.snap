// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessQueryPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 1`] = `
{
  "action": "delete",
  "principal": {
    "some": "principal",
  },
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
  },
}
`;

exports[`accessQueryPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 2`] = `
{
  "action": "create",
  "principal": {
    "some": "principal",
  },
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "organizationId",
      },
      "type": "organization",
    },
  },
}
`;

exports[`accessQueryPresetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 3`] = `
{
  "action": "update",
  "principal": {
    "some": "principal",
  },
  "subject": {
    "subjectType": "product",
    "wrappedSubject": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
  },
}
`;
